File: bindings/wasm/src/satox_bindings/network_manager.cpp
#include <emscripten/bind.h>
#include <string>

using namespace emscripten;

class NetworkManager {
public:
    NetworkManager() : initialized(false), connected(false), port(60777) {}

    void initialize() {
        if (initialized) throw std::runtime_error("NetworkManager already initialized");
        initialized = true;
    }

    bool isInitialized() const { return initialized; }

    void connect(const std::string& host, int port) {
        if (!initialized) throw std::runtime_error("NetworkManager not initialized");
        if (connected) throw std::runtime_error("Already connected");
        if (port != 60777) throw std::runtime_error("Invalid port for Satoxcoin network");
        connected = true;
        this->port = port;
    }

    void disconnect() {
        if (!connected) throw std::runtime_error("Not connected");
        connected = false;
    }

    bool isConnected() const { return connected; }
    int getPort() const { return port; }

private:
    bool initialized;
    bool connected;
    int port;
};

EMSCRIPTEN_BINDINGS(NetworkManager) {
    class_<NetworkManager>("NetworkManager")
        .constructor<>()
        .function("initialize", &NetworkManager::initialize)
        .function("isInitialized", &NetworkManager::isInitialized)
        .function("connect", &NetworkManager::connect)
        .function("disconnect", &NetworkManager::disconnect)
        .function("isConnected", &NetworkManager::isConnected)
        .function("getPort", &NetworkManager::getPort)
        ;
} 

File: bindings/typescript/src/managers/network-manager.ts
/**
 * @file network-manager.ts
 * @brief Network Manager for TypeScript/JavaScript bindings
 * @copyright Copyright (c) 2025 Satoxcoin Core Developers
 * @license MIT License
 */

import { EventEmitter } from 'events';
import { native } from '../../build/Release/satox_sdk_native';
import {
  NetworkConfig,
  IManager,
  SDKEvent,
  SatoxError,
  SatoxNetworkError,
  AsyncBoolean,
  AsyncString,
  AsyncNumber,
  AsyncVoid
} from '../types';

/**
 * Network Manager providing network functionality
 */
export class NetworkManager extends EventEmitter implements IManager {
  private _initialized = false;
  private _config: NetworkConfig = {};
  private _lastError = '';
  private _stats: Record<string, any> = {};

  constructor() {
    super();
    this.resetStats();
  }

  /**
   * Initialize the network manager
   */
  public async initialize(config: NetworkConfig): Promise<boolean> {
    try {
      if (this._initialized) {
        throw new SatoxNetworkError('Network manager already initialized');
      }

      this._config = { ...config };
      
      // Initialize native network manager
      const result = native.networkManagerInitialize(JSON.stringify(config));
      if (!result.success) {
        throw new SatoxNetworkError(result.error || 'Failed to initialize native network manager');
      }

      this._initialized = true;
      this.emit('initialized', { timestamp: new Date().toISOString() });

      return true;
    } catch (error) {
      this._lastError = error.message;
      const satoxError = error instanceof SatoxError ? error : new SatoxNetworkError(error.message);
      this.emit('error', satoxError);
      throw satoxError;
    }
  }

  /**
   * Shutdown the network manager
   */
  public async shutdown(): Promise<void> {
    try {
      if (!this._initialized) {
        return;
      }

      native.networkManagerShutdown();
      this._initialized = false;
      this.emit('shutdown', { timestamp: new Date().toISOString() });
    } catch (error) {
      this._lastError = error.message;
      const satoxError = error instanceof SatoxError ? error : new SatoxNetworkError(error.message);
      this.emit('error', satoxError);
      throw satoxError;
    }
  }

  /**
   * Check if network manager is initialized
   */
  public isInitialized(): boolean {
    return this._initialized;
  }

  /**
   * Get last error
   */
  public getLastError(): string {
    return this._lastError;
  }

  /**
   * Get statistics
   */
  public getStats(): Record<string, any> {
    return { ...this._stats };
  }

  /**
   * Reset statistics
   */
  public resetStats(): void {
    this._stats = {
      connections_established: 0,
      messages_sent: 0,
      messages_received: 0,
      last_operation: null
    };
  }

  /**
   * Health check for network manager
   */
  public async healthCheck(): Promise<boolean> {
    try {
      if (!this._initialized) {
        return false;
      }

      const result = native.networkManagerHealthCheck();
      return result.success;
    } catch {
      return false;
    }
  }
} 